File: Misc.vb
Imports System.IO
Imports System.IO.Compression
Imports BasicUFL
Public Module Misc
    Public Function Combine(ByVal arrays As Byte()()) As Byte()
        Dim ret As Byte() = New Byte(arrays.Sum(Function(x) x.Length) - 1) {}
        Dim offset As Integer = 0
        For Each data As Byte() In arrays
            Buffer.BlockCopy(data, 0, ret, offset, data.Length)
            offset += data.Length
        Next
        Return ret
    End Function
    Public Function TrimNull(ByVal Array As Byte()) As Byte()
        For i As Integer = Array.Length - 1 To 0 Step -1
            If Array(i) > 0 Then
                Return ToList(Array).GetRange(0, i + 1).ToArray()
            End If
        Next
        Return Nothing
    End Function
    Public Function Deflate(ByVal Bytes As Byte()) As Byte()
        Dim m As New MemoryStream()
        Dim d As New DeflateStream(m, CompressionMode.Compress)
        d.Write(Bytes, 0, Bytes.Length)
        d.Flush()
        d.Close()
        Return m.ToArray
    End Function
    Public Function Inflate(ByVal Bytes As Byte()) As Byte()
        Dim m As New MemoryStream(Bytes)
        Dim d As New DeflateStream(m, CompressionMode.Decompress)
        d.Write(Bytes, 0, Bytes.Length)
        d.Flush()
        d.Close()
        Return m.ToArray
    End Function
    Public Function Hash(ByVal account As String, ByVal password As String, ByVal challenge As String, ByVal rounds As Integer) As String
        Dim salt = System.Text.Encoding.UTF8.GetBytes(account & challenge)
        Dim sha = Security.Cryptography.SHA256.Create()
        Dim hsh As Byte() = sha.ComputeHash(System.Text.Encoding.UTF8.GetBytes(password))
        For i = 0 To rounds
            sha.Initialize()
            sha.TransformBlock(hsh, 0, hsh.Length, Nothing, 0)
            sha.TransformFinalBlock(salt, 0, salt.Length)
            hsh = sha.Hash
        Next
        Return Convert.ToBase64String(hsh)
    End Function
    Public Enum ChatChannel
        Universe = 1
        World = 0
        Whisper = 2
        CommandResult = 3
        White = 4
    End Enum
End Module

File: Misc.cs
using Microsoft.VisualBasic;
using System;
using Ionic.Zlib;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.IO;
using System.IO.Compression;
public static class Misc
{
	public static byte[] Combine(byte[][] arrays)
	{
		List<byte> ret = new List<byte>(); 
		foreach (byte[] data in arrays) {
            foreach (byte d in data) {
                ret.Add(d);
            }
		}
		return ret.ToArray();
	}
	public static byte[] TrimNull(byte[] Array)
	{
		for (int i = Array.Length - 1; i >= 0; i += -1) {
			if (Array[i] > 0) {
				return Misc.ToList(Array).GetRange(0, i + 1).ToArray();
			}
		}
		return null;
	}
	public static byte[] Deflate(byte[] Bytes)
	{
        return ZlibStream.UncompressBuffer(Bytes);
	}
	public static byte[] Inflate(byte[] Bytes)
	{
        return ZlibStream.CompressBuffer(Bytes);
	}
	public static string Hash(string account, string password, string challenge, int rounds)
	{
		dynamic salt = System.Text.Encoding.UTF8.GetBytes(account + challenge);
		dynamic sha = System.Security.Cryptography.SHA256.Create();
		byte[] hsh = sha.ComputeHash(System.Text.Encoding.UTF8.GetBytes(password));
		for (int i = 0; i <= rounds; i++) {
			sha.Initialize();
			sha.TransformBlock(hsh, 0, hsh.Length, null, 0);
			sha.TransformFinalBlock(salt, 0, salt.Length);
			hsh = sha.Hash;
		}
		return Convert.ToBase64String(hsh);
	}
    public static List<byte> ToList(byte[] lel)
    {
        return new List<byte>(lel);
    }
}
public enum ChatChannel
{
	Universe = 1,
   	World = 0,
	Whisper = 2,
	CommandResult = 3,
	White = 4
}
